File: apps/web/src/lib/random.ts
/**
 * Numerical Recipes LCG — deterministic `Math.random`-compatible [0, 1) stream.
 * Shared by tests and any caller that needs reproducible demo generation.
 */
export function createSeededRandom(seed: number): () => number {
  let state = seed;
  return () => {
    state = (state * 1_664_525 + 1_013_904_223) % 4_294_967_296;
    return state / 4_294_967_296;
  };
}

/**
 * Deterministic id factory derived from a numeric seed.
 * Used when `seed` is set without an explicit `idFactory` so the same seed
 * reproduces scenario / target / event ids (acceptance criterion #8).
 */
export function createSeededIdFactory(seed: number): () => string {
  let n = 0;
  const prefix = `s${(seed >>> 0).toString(36)}`;
  return () => `${prefix}-${(n++).toString(36)}`;
}

/** Resolve id factory: explicit override > seed-derived > crypto.randomUUID. */
export function resolveIdFactory(options: {
  seed?: number;
  idFactory?: () => string;
}): () => string {
  if (options.idFactory) return options.idFactory;
  if (options.seed !== undefined) return createSeededIdFactory(options.seed);
  return () => crypto.randomUUID();
}

File: apps/api/internal/geo/random.go
package geo

import "strconv"

// CreateSeededRandom returns the Numerical Recipes deterministic [0,1) LCG stream.
func CreateSeededRandom(seed uint32) func() float64 {
	state := seed
	return func() float64 {
		state = state*1664525 + 1013904223
		return float64(state) / 4294967296.0
	}
}

// CreateSeededIDFactory returns deterministic base-36 IDs derived from seed.
func CreateSeededIDFactory(seed uint32) func() string {
	prefix := "s" + strconv.FormatUint(uint64(seed), 36)
	var n uint64
	return func() string {
		id := prefix + "-" + strconv.FormatUint(n, 36)
		n++
		return id
	}
}
